### PR TITLE
Add matrix bias to forward/backward kernel

### PR DESCRIPTION
**Pull request by xinyazhang**
All UTs Passed. No performance regression on existing forward kernel cases (i.e., `test/performance_forward.py`).

### CODE REVIEW

File: tritonsrc/bwd_split_kernel.py [hunk 1]
```diff
@@ -102,6 +109,34 @@ def bwd_kernel_dk_dv(
...
+                block_shape=(BLOCK_M, BLOCK_N),
+                order=(1, 0)
+                )
+        if (stride_dbz == 0 or stride_dbh == 0) or stride_dbm == 0:
+            store_db = False
+        else:
+            store_db = True
+        # Still have to make one even if no_db = False
+        # due to a limit of Triton: runtime branches must have identical data types.
+        DB_block_ptr = tl.make_block_ptr(
```
Comment (groenenboomj) [inline tritonsrc/bwd_split_kernel.py:130]: I see a store but no load so in the end what is this storing?

**Review by groenenboomj — COMMENTED** [review]
I'm trusting the database changes but I was also concerned to not see testing above 2k sequence lengths as most use cases where we are beneficial are higher than that.

File: tritonsrc/bwd_split_kernel.py [hunk 2]
```diff
@@ -170,6 +212,24 @@ def bwd_kernel_dk_dv(
...
+            if q_padded and k_padded:  # CAVEAT: using "or" disables the partial boundary_check branches
+                bias = tl.load(B_block_ptr, boundary_check=(0,1), padding_option="zero")
+            elif q_padded:
+                bias = tl.load(B_block_ptr, boundary_check=(0,), padding_option="zero")
+            elif k_padded:
+                bias = tl.load(B_block_ptr, boundary_check=(1,), padding_option="zero")
+            else:
+                bias = tl.load(B_block_ptr)
+            """
+            bias = tl.load(B_block_ptr, boundary_check=(0,1), padding_option="zero")
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply tritonsrc/bwd_split_kernel.py:229]: I haven't check if it's a compiler problem, and even if it is we do not have the capacity to fix it.
The performance is not a concern before we can make it work.

File: tritonsrc/bwd_split_kernel.py (hunk 1 above)
Comment (xinyazhang) [reply tritonsrc/bwd_split_kernel.py:130]: D is the prefix for differentials, and DB is the differential of bias tensor.

**Comment by xinyazhang** [comment]
> I'm trusting the database changes but I was also concerned to not see testing above 2k sequence lengths as most use cases where we are beneficial are higher than that.

Any sequence longer than 2k will be treated as 2k when looking up the optimal kernel. I don't think the optimal compiler options will change for longer sequences.

File: tritonsrc/bwd_split_kernel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (groenenboomj) [reply tritonsrc/bwd_split_kernel.py:229]: Sounds good then.  Just make an issue so we don't lose track. We'll need to root cause this before we hit more boundary issues.

[the remark this answers is not in the document]
Comment (xinyazhang) [reply tritonsrc/bwd_split_kernel.py:229]: Filed as https://github.com/ROCm/aotriton/issues/17

File: tritonsrc/attn_torch_function.py [hunk 3]
```diff
@@ -474,6 +466,8 @@ def backward(ctx, do, _, fwd_tuning_result):
         else:
             BLOCK_M = 128
             BLOCK_N = 64
+        if q.dtype == torch.float32:
+            BLOCK_M //= 2
```
Comment (groenenboomj) [inline tritonsrc/attn_torch_function.py:470]: I take it this was from memory limitations?

Comment (xinyazhang) [reply tritonsrc/attn_torch_function.py:471]: Yes, but unfortunately this doesn't fix all problems.
fp32 gets broken in other places.

**Review by groenenboomj — APPROVED** [review]
Chatted offline, looks good.

**Pull request merged**